src/internal/_eqDeep.js

```diff
@@ -0,0 +1,64 @@
+var eq = require('../eq');
+var keys = require('../keys');
+var type = require('../type');
+
+// The algorithm used to handle cyclic structures is
+// inspired by underscore's isEqual
+module.exports = function _eqDeep(a, b, stackA, stackB) {
+    var typeA = type(a);
+    if (typeA !== type(b)) {
+        return false;
+    }
+
+    if (eq(a, b)) {
+        return true;
+    }
+
+    if (typeA == 'RegExp') {
+        // RegExp equality algorithm: http://stackoverflow.com/a/10776635
+        return (a.source === b.source) &&
+               (a.global === b.global) &&
+               (a.ignoreCase === b.ignoreCase) &&
+               (a.multiline === b.multiline);
+    }
+
+    var eqDate = typeA == 'Date';
+    if (eqDate || typeA == 'Object' || typeA == 'Array' || typeA == 'ArrayBuffer' ||
+        typeA == 'Int16Buffer' || typeA == 'Int8Array' || typeA == 'Uint8Array' ||
+        typeA == 'Uint8ClampedArray' || typeA == 'Int16Array' || typeA == 'Uint16Array' ||
+        typeA == 'Int32Array' || typeA == 'Uint32Array' || typeA == 'Float32Array' ||
+        typeA == 'Float64Array') {
```

Comment:
That's much cleaner, I think just ```isObject(a)``` should suffice as we check that the types are equal at line 9.